spatial/r3: add Triangle

soypat opened the pull request:
Please take a look.

Resolves part of #1791 
<!--
Checklist:

- API changes have been discussed
- code is goformated correctly (goimports)
- packages with generated code have had code generation run
- tests pass locally
- linked to relevant issues

Please make sure your commit message summary line and pull request
title match the Go convention; a one-line summary of the change,
prefixed by the primary affected package that should complete the
sentence, "This change modifies Gonum to _____."
-->

sbinet commented on spatial/r3/triangle.go:
```diff
@@ -0,0 +1,72 @@
+// is composed by 3 vectors corresponding to the position
+// of each of the vertices. Ordering of these vertices
+// decides the "normal" direction.
+type Triangle struct {
+	// Triangle vertices. Ordering of these vertices decides Normal vector direction.
+	// Inverting ordering of two vertices inverts the resulting direction.
+	V [3]Vec
+}
```
just a drive-by comment:
any reason why we don't define it as:
```go
type Triangle [3]Vec
```

or, (better for the SSA pass):
```go
type Triangle struct {
    A, B, C Vec
}
```


do we expect to sub-slice `t.V`? or iterate over the vertices of a triangle?

soypat replied:
It's common to iterate over triangle vertices to make code more succinct. I can change it to `type Triangle [3]Vec`, which I prefer over the current implementation.

sbinet commented on spatial/r3/triangle.go:
```diff
@@ -0,0 +1,72 @@
+// of the triangle vertices decides the normal's resulting
+// direction. Normal is not guaranteed to return a unit-norm vector.
+func (t Triangle) Normal() Vec {
+	s1, s2, _ := t.sides()
```
it seems (moderately) wasteful to ask for the computation of `Sub(t.V[0], t.V[2])` to drop it right there.
but ok.

soypat replied:
Would the SSA not optimize the calculation? Maybe not today but it does seem like something that could done given it looks like a short enough path

sbinet reviewed (approved):
just a little stream of consciousness.

LGTM, perhaps add a little example?

soypat commented:
@sbinet I've added an example for which I do not have the tools to validate. I have also added the `Centroid` method here and in the `r2` implementation.

sbinet commented:
although, it'd be great if you could rebase off #1801 (to get the CI running)

soypat commented:
@kortschak  I have added the tests requested with a handful of comments to help out readers. 

I managed to fix the icosphere example. Below are 0 and 7 subdivisions of an icosahedron.
![image](https://user-images.githubusercontent.com/26156425/168402482-3f9aa0aa-86ae-4dc5-8c08-1ea1d34a244b.png)
![image](https://user-images.githubusercontent.com/26156425/168402545-9064e4e5-f86a-4629-91d8-62fed9a648af.png)

kortschak commented on spatial/r3/triangle.go:
```diff
@@ -0,0 +1,116 @@
+
+// Degenerate returns true if all of triangle's vertices are
+// within tol distance of its longest side.
+func (t Triangle) Degenerate(tol float64) bool {
```
Consider renaming this to `IsDegenerate`.

kortschak commented on spatial/r3/triangle.go:
```diff
@@ -0,0 +1,116 @@
+
+// Area returns the surface area of the triangle.
+func (t Triangle) Area() float64 {
+	// Formula taken from Section 2, page 3.
+	// William M. Kahan (24 March 2000). "Miscalculating Area and Angles of a Needle-like Triangle"
+	// https://people.eecs.berkeley.edu/~wkahan/Triangle.pdf
```
Note that this is Heron's formula. I would probably link to the first (as well as this reference) given the ancient history of it. https://en.wikipedia.org/wiki/Heron%27s_formula.

```
// Heron's Formula, see https://en.wikipedia.org/wiki/Heron%27s_formula.
// Also see William M. Kahan (24 March 2000). "Miscalculating Area and Angles of a Needle-like Triangle"
// for more discussion. https://people.eecs.berkeley.edu/~wkahan/Triangle.pdf.
```

kortschak commented on spatial/r3/triangle.go:
```diff
@@ -0,0 +1,116 @@
+	return longIdx
+}
+
+// orderedLengths returns the lengths of the sides of the triangle such that
+// a ≤ b ≤ c.
+func (t Triangle) orderedLengths() (a, b, c float64) {
+	s1, s2, s3 := t.sides()
+	l1 := Norm(s1)
+	l2 := Norm(s2)
+	l3 := Norm(s3)
+	// sort-3
+	if l2 < l1 {
+		l1, l2 = l2, l1
+	}
+	if l3 < l2 {
+		l2, l3 = l3, l2
+		if l2 < l1 {
+			l1, l2 = l2, l1
+		}
+	}
+	return l1, l2, l3
+}
```
Used only once for `Area`, so place it immediately after that.

kortschak commented on spatial/r3/triangle.go:
```diff
@@ -0,0 +1,116 @@
+	return Sub(t[1], t[0]), Sub(t[2], t[1]), Sub(t[0], t[2])
+}
+
+// longIdx returns index of the longest side. The sides
+// of the triangles are are as follows:
+//  - Side 0 formed by vertices 0 and 1
+//  - Side 1 formed by vertices 1 and 2
+//  - Side 2 formed by vertices 0 and 2
+func (t Triangle) longIdx() int {
+	sides := [3]Vec{Sub(t[1], t[0]), Sub(t[2], t[1]), Sub(t[0], t[2])}
+	len2 := [3]float64{Norm2(sides[0]), Norm2(sides[1]), Norm2(sides[2])}
+	longLen := len2[0]
+	longIdx := 0
+	if len2[1] > longLen {
+		longLen = len2[1]
+		longIdx = 1
+	}
+	if len2[2] > longLen {
+		longIdx = 2
+	}
+	return longIdx
+}
```
Since this is only used once, for `Degenerate`, I'd place it immediately after that.

kortschak reviewed (commented):
This is looking good.

kortschak commented:
Can you rebase onto the current master to get the new linter.

soypat commented on spatial/r3/icosahedron_example_test.go:
```diff
@@ -0,0 +1,98 @@
+		faces = append(faces, face)
+	}
+	fmt.Println(faces)
+	// The 3D rendering of the icosphere is left as an exercise to the reader.
```
Left a small pointer-comment, good enough?

kortschak replied:
Oh, sorry, no. OOB. This comment for the Barnes-Hut example was how the galaxy example for gio was instigated. This doesn't need to be here.

soypat commented on spatial/r3/icosahedron_example_test.go:
```diff
@@ -0,0 +1,98 @@
+// the vertices defining the edge that will be subdivided. lookup is a
+// table of all newly generated vertices from previous calls to subdivideEdge
+// so as to not duplicate vertices.
+func subdivideEdge(lookup map[edgeIdx]int, vertices []r3.Vec, first, second int) (int, []r3.Vec) {
```
I reversed the order of the functions but left them before the example following suit of the euler angle example. Should I add them after the example or is this OK?

kortschak replied:
The Euler example is a special case; it was in that order because people know (or think they know) what Euler Angles are and want to use them, and it was written as a warning. This example has helpers that are not (AFAICS) something to be copy-pasted into their code.

soypat commented:
I rebased onto current master, which is [ci: install ragel to check rdf codegen](https://github.com/gonum/gonum/commit/84fd2bc2bbc4a322613c3e2e81647e3a9682bcc8)

kortschak commented:
You're right, sorry. The linter change hasn't merged yet.

soypat commented:
Refactored icosphere example:
- All triangle normals now point owtward from icosphere center
- Bug caused duplicate loose subdiving vertices has been fixed, also clearer return value
- Example has been moved to front of file.

kortschak commented:
It might be worth inspecting the example's mesh for defect. You can do this by exporting it as an STL and inspecting it with meshmixer. If you don't have code for STL export, you can use this https://github.com/kortschak/stl.

soypat commented:
Do not merge. I believe I found a bug with Area function.

soypat commented:
Nevermind, all is good. I've added more tests to `TestTriangleArea` and a new `TestTriangleOrderedLengths` test.

soypat commented:
Hey there peeps! I'm gonna go ahead and share something I'm really proud of: 
![image](https://user-images.githubusercontent.com/26156425/168938320-280fbff4-5e89-4d0c-8961-ddc8ad0d46d0.png)
A sphere .stl file (**left**) was taken aas input and a signed distance function generated from it (**right**) using the [`Transform` type](https://github.com/gonum/gonum/pull/1802) and a generic rewrite of `kdtree` (just for kicks). I'm looking forward to exposing this functionality with gonum's upcoming `r3.Triangle` as an argument! This would allow users of `sdf` to edit .stl files as if they were native `sdf.SDF3` objects

sbinet commented on spatial/r3/icosahedron_example_test.go:
```diff
@@ -0,0 +1,104 @@
+// refer to the edge that will be subdivided.
+// lookup is a table of all newly generated vertices from
+// previous calls to subdivideEdge so as to not duplicate vertices.
+func subdivideEdge(lookup map[edgeIdx]int, vertices []r3.Vec, first, second int) (int, []r3.Vec) {
```
perhaps also return the lookup table, to future proof any behaviour change in the guts of the `map` implementation?

kortschak replied:
I don't think that's necessary; maps can't change their reference value behaviour because of the compatibility promise.
